### PR TITLE
[RPC] Refactored sessions.

### PR DESCRIPTION
**Pull request by kobolog**
Trello: https://trello.com/c/95HRjhQa

* Sessions no longer depend on `boost::signals2`. The shutdown signal wasn't required in the first place, since under highload system will reuse fds fast enough to replace stale sessions in the engine's map, and for quiet times engines will now garbage collect stale sessions every 60 seconds.
* Sessions now log stuff. That means improvements over `uncaught_error` without any details about what error was not actually caught.
* Much more predictable dispatch discard policy. Note that `discard()` call may or may not be accompanied by `std::error_code`.

### CODE REVIEW

**Comment by kobolog** [comment]
@3Hren @antmat: review away!

File: src/session.cpp [hunk 1]
```diff
@@ -246,20 +231,42 @@ session_t::inject(const dispatch_ptr_t& dispatch) {
 
 void
 session_t::revoke(uint64_t channel_id) {
-    channels->erase(channel_id);
+    channels.apply([&](channel_map_t& mapping) {
+        auto it = mapping.find(channel_id);
+
+        // NOTE: Not sure if that can ever happen, but that's why people use asserts, right?
+        BOOST_ASSERT(it != mapping.end());
```
Comment (3Hren) [inline src/session.cpp]: We always disable asserts for some reasons. I suggest to drop this check https://github.com/cocaine/cocaine-core/blob/v0.12/include/cocaine/common.hpp#L36

File: src/session.cpp [hunk 2]
```diff
@@ -328,7 +335,7 @@ session_t::memory_pressure() const {
 
 std::string
 session_t::name() const {
```
Comment (3Hren) [inline src/session.cpp]: We can return optional here, instead of strange strings. What if we have service named "<none>"?

File: src/session.cpp [hunk 3]
```diff
@@ -151,83 +129,87 @@ session_t::push_action_t::operator()(std::shared_ptr<channel<tcp>> ptr) {
...
 
         if(lb == ub) {
             if(channel_id <= max_channel_id) {
                 // NOTE: Checking whether channel number is always higher than the previous channel
                 // number is similar to an infinite TIME_WAIT timeout for TCP sockets. It might be
-                // not the best approach, but since we have 2^64 possible channels, and it is a lot
-                // more than 2^16 ports for sockets, it is fit to avoid stray messages.
-                return;
+                // not the best approach, but since we have 2^64 possible channels it's good enough.
+                throw cocaine::error_t("specified channel id was revoked");
```
Comment (3Hren) [inline src/session.cpp]: So, what's the reason to drop channels accepted in non-ascending order.

File: src/session.cpp [hunk 4]
```diff
@@ -142,7 +120,7 @@ class session_t::push_action_t:
 };
 
 void
-session_t::push_action_t::operator()(std::shared_ptr<channel<tcp>> ptr) {
+session_t::push_action_t::operator()(const std::shared_ptr<channel<tcp>> ptr) {
```
Comment (3Hren) [inline src/session.cpp]: Why const?

File: src/engine.cpp [hunk 5]
```diff
@@ -37,43 +37,89 @@ using namespace asio::ip;
...
 }
 
-} // namespace
+execution_unit_t::execution_unit_t(context_t& context):
+    m_asio(new io_service()),
+    m_chamber(new io::chamber_t("core:asio", m_asio)),
+    m_cron(*m_asio)
+{
+    m_log = context.log("core:asio", {
+        attribute::make("engine", boost::lexical_cast<std::string>(m_chamber->thread_id()))
```
Comment (3Hren) [inline src/engine.cpp]: JFYI, Blackhole already has `tid` attribute for all log records.

Comment (3Hren) [inline src/engine.cpp]: And LWP.

File: src/engine.cpp [hunk 6]
```diff
@@ -37,43 +37,89 @@ using namespace asio::ip;
...
-        COCAINE_LOG_DEBUG(m_log, "engine waiting for outstanding operations to complete");
+    m_asio->post([this] {
+        COCAINE_LOG_DEBUG(m_log, "stopping engine");
 
-        // Asynchronously close the connections.
-        m_asio->post(detach_action_t{m_sessions});
-    }
+        for(auto it = m_sessions.begin(); it != m_sessions.end(); ++it) {
+            // Close the connections.
+            it->second->detach(std::error_code());
```
Comment (3Hren) [inline src/engine.cpp]: Why we detach sessions with none error code instead of `operation_aborted`?

File: src/session.cpp [hunk 7]
```diff
@@ -343,10 +350,8 @@ session_t::remote_endpoint() const {
         try {
             endpoint = ptr->socket->remote_endpoint();
         } catch(const std::system_error& e) {
-            // TODO: Log this.
+            // Ignore.
```
Comment (3Hren) [inline src/session.cpp]: Are you sure, that to silently eat an exception is the best idea?

File: src/session.cpp (hunk 1 above)
Comment (kobolog) [inline src/session.cpp]: Nah, let's keep it. I'd rather suggest make it on for CMAKE_BUILD_TYPE=Debug by default.

File: src/session.cpp (hunk 2 above)
Comment (kobolog) [inline src/session.cpp]: I've been thinking about it for some time, actually. Does BH have some special handling magic for optional attributes?

Comment (3Hren) [inline src/session.cpp]: Nope.

File: src/session.cpp (hunk 1 above)
Comment (3Hren) [inline src/session.cpp]: Agree.

File: src/session.cpp (hunk 3 above)
Comment (kobolog) [inline src/session.cpp]: That's the easiest way to keep track of revoked channels. Do you know why we want to keep track of revoked channels rather than reuse them?

File: src/session.cpp (hunk 4 above)
Comment (kobolog) [inline src/session.cpp]: Because we don't mutate the pointer here, only the pointee?

File: src/engine.cpp (hunk 5 above)
Comment (kobolog) [inline src/engine.cpp]: Cool. That's useless anyway, only to tell engines apart in debug logs somehow, which might be useless as well.

File: src/engine.cpp (hunk 6 above)
Comment (kobolog) [inline src/engine.cpp]: I'm actually not sure about the error code here. It's not really an ERROR when we shutdown the engine normally as a part of runtime shutdown, right? Also, no operation was aborted, technically.

File: src/session.cpp (hunk 7 above)
Comment (kobolog) [inline src/session.cpp]: I'm not sure, but I guess in this case it's okay since there's nothing you can really do or fix or whatever about it.

File: src/session.cpp (hunk 1 above)
Comment (kobolog) [inline src/session.cpp]: Can you do it in a separate PR, since you know CMake better than me?

File: src/session.cpp (hunk 2 above)
Comment (kobolog) [inline src/session.cpp]: So that would make logging of such return values a mess. We'd either end up doing something like `session.name().get_value_or("<none>")` all the same or writing a separate `COCAINE_LOG_*` for each case.

File: src/session.cpp (hunk 3 above)
Comment (3Hren) [inline src/session.cpp]: To prevent repeatedly handling duplicate packages, I can only guess.

File: src/session.cpp (hunk 1 above)
Comment (3Hren) [inline src/session.cpp]: Okay, someday =)

File: src/session.cpp (hunk 3 above)
Comment (kobolog) [inline src/session.cpp]: Actually, to prevent a situation when a client sends messages to a channel ignoring all errors coming back. Imagine a situation where a client sends `storage::write_streamed(collection, key)` followed by a number of `streaming::write` invocations. In case of the former method actually failing and returning an error, our clueless client would end up calling some completely random method instead of 'streaming::write' since the channel was reset after the error.

Comment (3Hren) [inline src/session.cpp]: Okay, got it.
I think it'll be cool to copy this description directly into the code to prevent further misunderstanding.
